include/rspq.h

```diff
@@ -0,0 +1,704 @@
+ * command that will call A.
+ *
+ * @param block The block that must be run
+ * 
+ * @note The maximum depth of nested block calls is 8.
+ */
+void rspq_block_run(rspq_block_t *block);
+
+/**
+ * @brief Free a block that is not needed any more.
+ * 
+ * After calling this function, the block is invalid and must not be called
+ * anymore.
+ * 
+ * @param  block  The block
+ * 
+ * @note If the block was being called by other blocks, these other blocks
+ *       become invalid and will make the RSP crash if called. Make sure
+ *       that freeing a block is only done when no other blocks reference it.
+ */
+void rspq_block_free(rspq_block_t *block);
+
+/**
+ * @brief Start building a high-priority queue.
+ * 
+ * This function enters a special mode in which a high-priority queue is
+ * activated and can be filled with commands. After this function has been
+ * called, all commands will be put in the high-priority queue, until
+ * #rspq_highpri_end is called.
+ * 
+ * The RSP will start processing the high-priority queue almost instantly
+ * (as soon as the current command is done), pausing the normal queue. This will
+ * also happen while the high-priority queue is being built, to achieve the
+ * lowest possible latency. When the RSP finishes processing the high priority
+ * queue (after #rspq_highpri_end closes it), it resumes processing the normal
+ * queue from the exact point that was left.
+ * 
+ * The goal of the high-priority queue is to either schedule latency-sensitive
+ * commands like audio processing, or to schedule immediate RSP calculations
+ * that should be performed right away, just like they were preempting what
+ * the RSP is currently doing.
+ * 
+ * It is possible to create multiple high-priority queues by calling
+ * #rspq_highpri_begin / #rspq_highpri_end multiples time with short
+ * delays in-between. The RSP will process them in order. Notice that
+ * there is a overhead in doing so, so it might be advisable to keep
+ * the high-priority mode active for a longer period if possible. On the
+ * other hand, a shorter high-priority queue allows for the RSP to
+ * switch back to processing the normal queue before the next one
+ * is created.
+ * 
+ * @note It is not possible to create a block while the high-priority queue is
+ *       active. Arrange for constructing blocks beforehand.
+ *       
+ * @note It is currently not possible to call a block from the
+ *       high-priority queue. (FIXME: to be implemented)
+ *       
+ */
+void rspq_highpri_begin(void);
+
+/**
+ * @brief Finish building the high-priority queue and close it.
+ * 
+ * This function terminates and closes the high-priority queue. After this
+ * command is called, all following commands will be added to the normal queue.
+ * 
+ * Notice that the RSP does not wait for this function to be called: it will
+ * start running the high-priority queue as soon as possible, even while it is
+ * being built.
```

Comment:
Does this also mean we don't need `rspq_flush` when inside a `rspq_highpri_begin`?
Something to keep in mind regarding multithreading: multiple threads might call this before one ended the high prio queue. I know we don't support re-entrancy for rspq ATM but it looked like a little trickier than other cases here.

A question: will this cause a bottleneck b/c CPU is unable to push enough commands to the high priority queue? OTOH I don't think it will ever be an issue because each command will take enough time on RSP that CPU will probably never lag behind.

Reply:
> A question: will this cause a bottleneck b/c CPU is unable to push enough commands to the high priority queue? OTOH I don't think it will ever be an issue because each command will take enough time on RSP that CPU will probably never lag behind.

Yeah I mean it could cause a small delay on RSP, but it's the price to pay if you have high priority aka low latency. I wouldn't say a bottleneck in general. If the CPU needs to write 1 highpri command, then do a long calculation, then write another highpri command, it can safely turn off highpri mode after the first command and turn it back again when the second command is ready to be written. What happens in this case is that if the RSP was still executing the first command, it will then smoothly go to the second command before going back to lowpri queue. Otherwise, if the RSP had already finished the first command and get back to lowpri, it will have to switch again to highpri. So it's up to the caller to decide whether use a single longer highpri mode (less latency, but possible RSP idle time) or two smaller highpri modes (potentially more latency, but no RSP idle time).